File: string.c
#include <stdio.h>  
#include <stdlib.h> 
#include<string.h>
#include<ctype.h>
void words(char *,const char *,char *); //поиск самого длинного слова, реализация перебором всех символов строки
void words(char *s,const char *q,char *word){//s -- строка, в которой выделяем слова, q -- набор разделителей, word -- буфер для сохранения самого длинного слова в строке
	unsigned int stat=0;//stat -- переменная, значение которой определяет статус символа в строке (находится он внутри или вне слова)
	while(1){ //цикл для перебора всех символов строки
		if(strchr(q,*s)){ //если очередной символ строки является разделителем
			if(stat>strlen(word)){//если *s -- это первый разделитель после слова, сравниваем длину предыдущего слова (длина которого накоплена в stat) с длиной слова, хранящегося в word
				strncpy(word,s-stat,stat);//если выделенное слово большей длины, копируем его в word
				word[stat]=0;//strncpy не выставляет конец строки, делаем это вручную
			}
			stat=0;//т.к. текущий символ является разделителем, обнуляем stat
		}
		else
			stat++; //т.к.  находимся внутри слова, увеличиваем количество составляющих слово букв
		if(*s==0) // если просмотр строки закончен
			break;//выходим из цикла
		s++;
	}
}
void search(const char *SInputFile, char *word); //поиск самого длинного слова в файле согласно набору разделителей  ,, .\t\n, использующий функцию words
void search(const char *SInputFile, char *word){
	FILE *in=fopen(SInputFile,"r");//открываем файл с заданным именем на чтение
	char buf[111]; //в этот буфер будем прочитывать строку файла
	word[0]=0;//начинаем со слова минимальной длины
	while(fgets(buf,111,in)!=NULL) //пока в файле остались символы, записываем очередную строку в буфер, при этом  текущее положение указателя внутри файла смещается к началу следующей строки
		words(buf,",, .\t\n",word);//режем строку по заданному набору разделителей и слово максимальной длины сохраняем в word
	fclose(in);
}
void search1(const char *SInputFile, char *word); //поиск самого длинного слова в файле согласно набору разделителей  ,, .\t\n, использующий стандартную функцию strtok
void search1(const char *SInputFile, char *word){
	FILE *in=fopen(SInputFile,"r");//открываем файл с заданным именем на чтение
	char buf[111]; //в этот буфер будем прочитывать строку файла
	word[0]=0;//начинаем со слова минимальной длины
	while(fgets(buf,111,in)!=NULL) //пока в файле остались символы, записываем очередную строку в буфер, при этом  текущее положение указателя внутри файла смещается к началу следующей строки
	for(char *begin=strtok(buf,",, .\t\n");begin;begin=strtok(NULL,",,. \t\n")) //перебираем слова строки buf
		if(strlen(begin)>strlen(word)) //если нашлось более длинное слово, чем word
			strcpy(word,begin); //обновляем word
	fclose(in);
}
void search2(const char *SInputFile, char *word);//поиск самого длинного слова файла, здесь под словом понимается участок строки между символами "
void search2(const char *SInputFile, char *word){
	FILE *in=fopen(SInputFile,"r");
	char buf[111],*begin,*end;
	memset(word,0,111);//это избавит нас от необходимости прописывать конец строки
	while(fgets(buf,111,in)!=NULL)
	for((begin=strchr(buf,'\"'),begin?end=strchr(begin+1,'\"'):NULL) //находим в строке, если это возможно, пару символов "
		;begin; //for будет работать, когда такая пара кавычек нашлась (т.е. можно выделить очередное слово в строке)
		(begin=strchr(end+1,'\"'),begin?end=strchr(begin+1,'\"'):NULL))//на следующей итерации for передвигаемся по строке и пытаемся выделить следующее слово
		if(end-begin-2>(int)strlen(word)) //если найденное слово длиннее word
			strncpy(word,begin+1,end-begin-1); //копируем его
	fclose(in);
}

int main(void){
	char word[111];
	char s[]="   yr\ny  g\t    ghghg    ";
	int k;
	for(char *begin=s;sscanf(begin,"%s%n",word,&k)==1;begin+=k) //выделение слов из строки с использованием sscanf
		puts(word);                //работает только для пробельных разделителей
        search2("1.txt",word);
	puts(word);
	return 0;
}
 

File: string.py
def words(s,q,word):# s -- строка, в которой выделяем слова, q -- набор разделителей, возвращает первое самое длинное слово в строке s согласно набору разделителй среди слов, длинне чем word
	stat=0   #в этой переменной сохраняется длина текущего слова
	i=0  #тндекс символа в строке
	while(True):# /цикл для перебора всех символов строки
		if(s[i] in q):# если очередной символ строки является разделителем
			if(stat>len(word)):#если s[i] -- это первый разделитель после слова, сравниваем длину предыдущего слова (длина которого накоплена в stat) с длиной слова, хранящегося в word
				word=s[i-stat:i]#если выделенное слово имеет длину, ольшую, чем word, вырезаем его из s и заменяем word на это слово
			stat=0 # дляследующего слова
		else:
			stat+=1  # если находимся внутри слова (текущий символ не является разделителем), увеличиваем длину
		i+=1 #переходим к следующему имволу строки s
		if(i==len(s)): #если символ s[i-1] был последним в строке s
			if stat>len(word): #если последнее слово имее длину, большую, чем wosd
				word=s[i-stat:i]
			break#//выходим из цикла
	return word

def search(SInputFile, q):
	f=open(SInputFile,"r")
	word=""
	for buf in f:
		word=words(buf,q,word)
	f.close()
	return word

def main():
	s="  22  yr\ny  g\t 2222222   wghghg5p5"
	print("Begin",s)
	word=words(s," \t\nw","")
	print('Resultfrom string  ',word,'len=',len(word))
	word=search("1.txt"," \t\n")
	print('Result from file',word,'len=',len(word))
	return 0

main() 